components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt

```diff
@@ -303,6 +308,12 @@ class DownloadsFeature(
         return findPreviousAppDownloaderDialogFragment() != null
     }
 
+    internal fun closeDownloadResponse(tabId: String) {
+        store.state.findTabOrCustomTab(tabId)?.let {
+            it.content.download?.response?.close()
```

Comment:
This is another interesting case where we have state in the store and we apply a side effect on it. Ideally, we would do this in middlewares only, but no reason to block this right now.

Reply:
Alternatively, maybe `consumeDownload` should close the response?

Reply:
> Alternatively, maybe consumeDownload should close the response?


That would be ideal but unfortunately not always when we `consumeDownload` we have to close the download stream, 
this only happens when the download is canceled. As the action of consuming the download just removes the [download request](https://github.com/mozilla-mobile/android-components/blob/8a55899f2fca7983d01c6e2515d53e26b500765d/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/ContentStateReducer.kt#L73) from the session and we need the download stream opened for later performing the download, because after a download is started we consume the download request and we [transfer](https://github.com/mozilla-mobile/android-components/blob/8a55899f2fca7983d01c6e2515d53e26b500765d/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/DownloadStateReducer.kt#L18) the download to the [BrowserState.downloads](https://github.com/mozilla-mobile/android-components/blob/8a55899f2fca7983d01c6e2515d53e26b500765d/components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt#L39)  **using the same download object** with the same `response` stream.   



I can create separate action and put the behavior on the `DownloadMiddleware`, what do you think?

Reply:
Hm, true, maybe a `CancelDownloadAction` plus a corresponding use case that dispatches it? This way the API would be symmetric and you could move the side effect to the middleware? That seems right to me, but no need to block in my opinion.

Reply:
👍🏽 
I will time-box it,if it take too much time I will address it in a separate ticket, thanks for the feedback.

Reply:
I will address it as a follow up to allow the patch to land https://github.com/mozilla-mobile/android-components/issues/9806